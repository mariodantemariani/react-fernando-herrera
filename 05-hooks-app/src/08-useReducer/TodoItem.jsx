import PropTypes from "prop-types";

export const TodoItem = ({ todo, onDeleteTodo, onToggleTodo }) => {
  //  const { id, description } = todo;

  return (
    <li
      key={todo.id}
      className={`list-group-item d-flex justify-content-between`}
    >
      <span
        className={`align-self-center ${
          todo?.done && "text-decoration-line-through"
        }`}
        onClick={() => onToggleTodo(todo.id)}
      >
        {todo.description}
      </span>
      <button
        className="btn btn-danger align-self-center btn-sm"
        onClick={() => onDeleteTodo(todo.id)}
      >
        Delete
      </button>
    </li>
  );
};

TodoItem.propTypes = {
  todo: PropTypes.object,
  //onDeleteTodo: PropTypes.func,
  //onTogggleTodo: PropTypes.func,
};
