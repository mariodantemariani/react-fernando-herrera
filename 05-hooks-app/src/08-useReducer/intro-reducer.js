// const initialState = [
//   {
//     id: 1,
//     todo: "Recolectar la piedra del Alma",
//     done: false,
//   },
//   {
//     id: 2,
//     todo: "Recolectar la piedra del Cielo",
//     done: false,
//   },
//   {
//     id: 3,
//     todo: "Recolectar la piedra del Espacio",
//     done: false,
//   },
// ];

// const todoReducer = (state = initialState, action = {}) => {
//   if (action?.type === "[TODO] add todo") {
//     return [...state, action.payload];
//   }
// };

// let todos = todoReducer();

// const newTodo = {
//   id: 4,
//   todo: "Recolectar la piedra del Tiempo",
//   done: false,
// };

// const addTodoAction = {
//   type: "[TODO] add todo",
//   payload: newTodo,
// };

// // todos = todoReducer(todos, addTodoAction);

// // console.log({ todos });
