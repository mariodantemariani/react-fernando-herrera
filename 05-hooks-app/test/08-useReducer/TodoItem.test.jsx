import { render, screen, fireEvent } from "@testing-library/react";
import { TodoItem } from "../../src/08-useReducer/TodoItem";

describe("Pruebas en TodoItem", () => {
  const todo = {
    id: 1,
    description: "Demo todo 1",
    done: false,
  };
  const onDeleteTodoMock = jest.fn();
  const onToggleTodoMock = jest.fn();

  beforeEach(() => jest.clearAllMocks());

  test("should be the same that snapShot", () => {
    const { container } = render(
      <TodoItem
        todo={todo}
        onDeleteTodo={onDeleteTodoMock}
        onToggleTodo={onToggleTodoMock}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test("should render Todo pending to complete", () => {
    render(
      <TodoItem
        todo={todo}
        onDeleteTodo={onDeleteTodoMock}
        onToggleTodo={onToggleTodoMock}
      />
    );

    const liElement = screen.getByRole("listitem");

    expect(liElement.className).toBe(
      "list-group-item d-flex justify-content-between"
    );

    const spanElement = screen.getByLabelText("span");

    expect(spanElement.className).toContain("align-self-center");
    expect(spanElement.className).not.toContain("text-decoration-line-through");
  });

  test("should render Todo completed", () => {
    todo.done = true;
    render(
      <TodoItem
        todo={todo}
        onDeleteTodo={onDeleteTodoMock}
        onToggleTodo={onToggleTodoMock}
      />
    );

    const liElement = screen.getByRole("listitem");

    expect(liElement.className).toBe(
      "list-group-item d-flex justify-content-between"
    );

    const spanElement = screen.getByLabelText("span");

    expect(spanElement.className).toContain("text-decoration-line-through");
  });

  test("should call onToggleTodo when render TODO completed", () => {
    todo.done = false;
    render(
      <TodoItem
        todo={todo}
        onDeleteTodo={onDeleteTodoMock}
        onToggleTodo={onToggleTodoMock}
      />
    );

    const spanElement = screen.getByLabelText("span");
    fireEvent.click(spanElement);

    expect(onToggleTodoMock).toHaveBeenCalledWith(todo.id);
  });

  test("should call onDelete", () => {
    render(
      <TodoItem
        todo={todo}
        onDeleteTodo={onDeleteTodoMock}
        onToggleTodo={onToggleTodoMock}
      />
    );

    const deleteButton = screen.getByRole("button");
    fireEvent.click(deleteButton);

    expect(onDeleteTodoMock).toHaveBeenCalledWith(todo.id);
  });
});
