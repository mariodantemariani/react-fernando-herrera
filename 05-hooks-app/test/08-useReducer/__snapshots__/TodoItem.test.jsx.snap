// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoItem should be the same that snapShot 1`] = `
<div>
  <li
    class="list-group-item d-flex justify-content-between"
  >
    <span
      aria-label="span"
      class="align-self-center false"
    >
      Demo todo 1
    </span>
    <button
      aria-label="button"
      class="btn btn-danger align-self-center btn-sm"
    >
      Delete
    </button>
  </li>
</div>
`;
